### PR TITLE
Complete() and Then() are verifiably reentrant

### LINKED ISSUE #1: Then() may not be reentered from Complete()
The following cod results in a deadlock.

```
	a := Promise()
	b := Promise()

	b.Combine(func(value interface{}) Thenable {
		bValue, _ := value.(int)

		return a.Then(func(value interface{}) interface{} {
			aValue, _ := value.(int)

			return bValue + aValue
		})
	}).Then(func(value interface{}) interface{} {
		sumValue, _ := value.(int)

		if sumValue != 5 {
			test.Fatalf("Expected sumValue(%d) to be 5", sumValue)
		}

		return nil
	})

	// This is the strange edge case where this occurs.  For some reason
	// calling complete on promise A must cause Then() (or similar) on A to be
	// called.
	a.Then(func(value interface{}) interface{} {
		b.Complete(3)

		return nil
	})

	a.Complete(2)
```

This is because `Then()` may not be invoked on a promise from within the callback chain executed during `Complete()`, due to the fact that `Complete()` holds a mutex until the end of the execution while `Then()` also attempts to acquire that same mutex, and these mutexes are not reentrant.

### PR DESCRIPTION
**Pull request by ssmccoy**
A deadlock was discovered in the case where `Then()` is called in the
completion chain on a promise which is presently being satisified with
`Complete()`. This has the opportunity to occur naturally in some
software with very complex workflows.

Included a test which verifiably forces this deadlock, which was used
to validate the fix, as well as a more complex test which simply
hammers a single promise concurrently in many goroutines to ensure
that it behaves consistently.

Resolves: #1

### CODE REVIEW

File: completable.go [hunk 1]
```diff
@@ -160,16 +169,25 @@ func (promise *CompletablePromise) Complete(value interface{}) {
 		panicStateComplete(promise.state == REJECTED)
 	}
 
-	composed := value
-
-	if promise.compute != nil {
-		composed = promise.compute(value)
-	}
-
```
Comment (sfllaw) [inline completable.go:168]: Yes, the mutex should only protect the mutation of the data structure.

**Review by sfllaw — APPROVED** [review]
